### DISCUSSION TITLE
Enabling FastAPI routing to functions

### DISCUSSION THREAD

**Discussion by jtrungle** [writer/writer-framework#34] [General]
Hi, first of all I would like to add my personal feedback -  this is an awesome project and definitely think this will change the game. 

I am a data analyst dabbling in development and have been building a lot in Streamlit applications for data enablement in my job.  This project definitely solves all of my gripes of Streamlit.

I was wondering, given that this is built off of FastAPI, is there currently any way to add routes to particular event/internal functions to expose as endpoints of the API? And by extension, could I hypothetically provide pydantic data validation in the response model?

Mainly, I'd like to expose some of the crud actions which would be useful for external actions via http request 

Otherwise - keen to see how this project progresses.
Thanks!
Johnny

**Comment by ramedina86** [comment]
Hi, thanks for the kind words, I'm really glad to hear you like Streamsync and see its potential.

The challenge that I see with exposing functions as FastAPI routes is linked to `state` and `session` management. How would you know that it's John's state and not Jane's state you're interested in? Or would you be interested in exposing something that's independent from session and state?

e.g. consider:

```py
def handle_click(state):
  state["counter"] += 1
```

If I make a POST request to `/event_handlers/handle_click`... whose counter do I increment?

Please let me know more about the use case you have in mind, I'm quite curious, sounds like it'd be interesting functionality.

PS: If you're interested in running FastAPI routes and Streamsync on the same ASGI server, check out the [Custom Server](https://www.streamsync.cloud/custom-server.html) docs. However, here the other routes stay completely independent from Streamsync, probably not what you're after?

**Reply by jtrungle** [reply]
Hi, thanks for the pointer on the custom server docs. I think creating the routes on the custom ASGI is probably what I was after. It would be better to separate the functions and call it separately from Streamsync code and the ASGI server as a route.
Jumped the gun a little with this question and was entertaining the possibilities for ease of development and maximise the utility of FastAPI. Thanks :)
